### DISCUSSION TITLE
感觉不支持udp

### DISCUSSION THREAD

**Discussion by yiguihai11** [xjasonlyu/tun2socks#120] [General]
我本地机器用ss客户端监听了socks5端口然后让tun2socks把流量发送到这个端口tcp正常udp死活不通，然后我又用tun2socks自带的ss代理udp就正常了。

**Comment by xjasonlyu** [comment]
你ss客户端开了UDP了吗

**Reply by yiguihai11** [reply]
> 你ss客户端开了UDP了吗

开了。客户端有监听udp端口服务端也支持udp，其它的像ssr,v2ray之类的也是不通。我再重复详细测试一遍把流程截图给你看下吧

**Reply by xjasonlyu** [reply]
SOCKS5 本身的协议实现有时在有多个内网的情况下会有不通的情况，你可以抓包试下，也有可能是因为你ss客户端的实现？

**Reply by yiguihai11** [reply]
> SOCKS5 本身的协议实现有时在有多个内网的情况下会有不通的情况，你可以抓包试下，也有可能是因为你ss客户端的实现？

这个是ss-rust版本客户端日志
```
root@BoringClever-VM:~# /etc/ssmanager/usr/bin/sslocal --server-url "ss://YWVzLTI1Ni1nY206UmV4bkJnVTdFVjVBRHhH@149.202.82.172:7002#%E5%85%AC%E7%9B%8A%E6%9C%BA%E5%9C%BA%3Ahttps%3A%2F%2Fbit.ly%2F3BPeo5G" --log-config /root/log4rs.yaml --log-without-time --local-addr 127.0.0.1:1080 -U
2022-04-18T23:25:30.732779371+08:00 DEBUG rlimit NOFILE adjusted rlimit { rlim_cur: 1048576, rlim_max: 1048576 }
2022-04-18T23:25:30.734079889+08:00 INFO  shadowsocks local 1.14.4 build 2022-04-10T00:11:59.172726063+00:00
2022-04-18T23:25:30.736286941+08:00 INFO  shadowsocks socks TCP listening on 127.0.0.1:1080
2022-04-18T23:25:30.737047560+08:00 INFO  shadowsocks socks5 UDP listening on 127.0.0.1:1080
2022-04-18T23:25:50.938048154+08:00 DEBUG CONNECT 104.16.132.229:443
2022-04-18T23:25:51.029084176+08:00 DEBUG established tcp tunnel 127.0.0.1:34120 <-> 104.16.132.229:443 through sever 149.202.82.172:7002 (outbound: 149.202.82.172:7002)
2022-04-18T23:26:16.257539044+08:00 DEBUG UDP ASSOCIATE from 0.0.0.0:0
2022-04-18T23:26:17.743696088+08:00 DEBUG UDP ASSOCIATE from 0.0.0.0:0
```
这是tun2socks启动参数
```
tun2socks --device tun://tun0 --proxy socks5://127.0.0.1:1080 -loglevel debug -interface eth0
```
日志
```
time="2022-04-18T23:23:05+08:00" level=info msg="[DIALER] bind to interface: eth0"
time="2022-04-18T23:23:05+08:00" level=info msg="[STACK] tun://tun0 <-> socks5://127.0.0.1:1080"
time="2022-04-18T23:23:06+08:00" level=info msg="[TCP] 172.19.0.1:58590 <-> 104.16.132.229:443"
time="2022-04-18T23:23:07+08:00" level=info msg="[TCP] 172.19.0.1:56820 <-> 172.67.22.102:443"
time="2022-04-18T23:23:08+08:00" level=info msg="[UDP] 172.19.0.1:35096 <-> 104.16.132.229:443"
time="2022-04-18T23:23:09+08:00" level=info msg="[UDP] 172.19.0.1:50150 <-> 104.16.133.229:443"
time="2022-04-18T23:23:31+08:00" level=info msg="[TCP] 172.19.0.1:58592 <-> 104.16.132.229:443"
time="2022-04-18T23:25:51+08:00" level=info msg="[TCP] 172.19.0.1:58594 <-> 104.16.132.229:443"
time="2022-04-18T23:26:16+08:00" level=info msg="[UDP] 172.19.0.1:34003 <-> 104.16.132.229:443"
time="2022-04-18T23:26:17+08:00" level=info msg="[UDP] 172.19.0.1:51118 <-> 104.16.133.229:443"
```
curl测试方法
```
curl --output /dev/null --write-out '%{response_code}' --connect-timeout 3 --max-time 3 --http3 --interface tun0 -4 --url https://cp.cloudflare.com/generate_204
```
使用curl测试tcp通udp不通

**Reply by xjasonlyu** [reply]
wireshark 或者 tcpdump 抓一下？目前的日志基本看不出问题

**Reply by yiguihai11** [reply]
> wireshark 或者 tcpdump 抓一下？目前的日志基本看不出问题

使用tcpdump抓包
```
tcpdump -n -i tun0 -vv -w /tmp/xxx.cap
```
[xxx.zip](https://github.com/xjasonlyu/tun2socks/files/8509009/xxx.zip)
这是抓包文件我分别测试了tcp和udp您帮忙分析一下

**Reply by xjasonlyu** [reply]
你要抓至少tun0和eth0两张网卡的才行

**Reply by yiguihai11** [reply]
> 你要抓至少tun0和eth0两张网卡的才行

这应该可以了吧
```
tcpdump -n -vv -w /tmp/xxx.cap
```
[xxx.zip](https://github.com/xjasonlyu/tun2socks/files/8509048/xxx.zip)

**Reply by yiguihai11** [reply]
> 你要抓至少tun0和eth0两张网卡的才行

_Originally posted by @zonyitoo in https://github.com/shadowsocks/shadowsocks-rust/discussions/816#discussioncomment-2590190_

**Reply by xjasonlyu** [reply]
看了一下抓包的数据，UDP包是发出去了，但没有接收到回复

**Reply by zonyitoo** [reply]
@yiguihai11 Add `-vvv` on `sslocal` could see that `trace!` log about receiving packets.

I am highly suspect that the packet was received by `sslocal`, but the packets that sent by it was somehow lost.

You can first verify the SOCKS5 UDP is working properly in your environment with the udpcheck.py script.

**Reply by xjasonlyu** [reply]
Agree, I just tested that socks5 is working well between sslocal and tun2socks.

**Reply by yiguihai11** [reply]
系统重装过后情况还是一样，我把自己的步骤写下来吧，完全可以自己重现。   
udp正常情况curl返回204响应码证明curl测试方法没有问题
```
直观点打开三个终端窗口分别执行
1窗
tun2socks --device tun://tun0 --proxy "ss://YWVzLTI1Ni1nY206UmV4bkJnVTdFVjVBRHhH@149.202.82.172:7002#%E5%85%AC%E7%9B%8A%E6%9C%BA%E5%9C%BA%3Ahttps%3A%2F%2Fbit.ly%2F3BPeo5G" -loglevel debug -interface eth0
2窗
ip -4 addr add 172.19.0.1/30 dev tun0
ip link set dev tun0 up
3窗
curl --output /dev/null --write-out '%{response_code}' --connect-timeout 3 --max-time 3 --http3 --interface tun0 -4 --url https://cp.cloudflare.com/generate_204
```  

重头戏来了把ss://替换成socks5:// 发送到sslocal
```
1窗
tun2socks --device tun://tun0 --proxy socks5://127.0.0.1:1080 -loglevel debug -interface eth0
2窗
ip -4 addr add 172.19.0.1/30 dev tun0
ip link set dev tun0 up
3窗
sslocal --server-url "ss://YWVzLTI1Ni1nY206UmV4bkJnVTdFVjVBRHhH@149.202.82.172:7002#%E5%85%AC%E7%9B%8A%E6%9C%BA%E5%9C%BA%3Ahttps%3A%2F%2Fbit.ly%2F3BPeo5G" --local-addr 127.0.0.1:1080 -U
4窗
curl --output /dev/null --write-out '%{response_code}' --connect-timeout 3 --max-time 3 --http3 --interface tun0 -4 --url https://cp.cloudflare.com/generate_204
```
一样的ss链接账号但是却没有正常的响应
[curl.zip](https://github.com/xjasonlyu/tun2socks/files/8516044/curl.zip)
如果操作步骤有错劳烦指出

**Comment by zonyitoo** [comment]
If you want to verify whether the SOCKS5 UDP relay is working properly, you can simply use this script:

[udpchk.py.zip](https://github.com/xjasonlyu/tun2socks/files/8510906/udpchk.py.zip)

It will send a DNS query to `114.114.114.114.53` through the SOCKS5 proxy, with Python's `socks` library (standard):

```
python udpcheck.py --proxy 127.0.0.1 --port 1080
```

**Reply by yiguihai11** [reply]
> If you want to verify whether the SOCKS5 UDP relay is working properly, you can simply use this script:
> 
> [udpchk.py.zip](https://github.com/xjasonlyu/tun2socks/files/8510906/udpchk.py.zip)
> 
> It will send a DNS query to `114.114.114.114.53` through the SOCKS5 proxy, with Python's `socks` library (standard):
> 
> ```
> python udpcheck.py --proxy 127.0.0.1 --port 1080
> ```

![IMG_20220419_161252](https://user-images.githubusercontent.com/103910817/163956951-5f5c4273-4031-40b8-954f-459f663f0c0c.jpg)
这个脚本应该不太行了一个不存在的端口都能显示pass  
不是curl测试方法的问题我用tun2socks自带的ss模块和sslocal的tun模式测试都可以，但是用tun2socks的socks5方式就是不行。
tun2socks的日志也显示  [UDP] 172.19.0.1:51118 <-> 104.16.133.229:443  发送去这个地址了。但是sslocal却显示 from 0.0.0.0:0 这就是我纳闷的原因

**Reply by zonyitoo** [reply]
The `0.0.0.0:0` is not related to this problem, JUST IGNORE IT!

Why in your environment every port can handle SOCKS5 UDP ASSOCIATE?? That's not expected!

If the proxy address is not listened, then it should print:

```
error(39, 'Destination address required')
error(39, 'Destination address required')
error(39, 'Destination address required')
error(39, 'Destination address required')
...
```

**Reply by xjasonlyu** [reply]
感觉这是配置了什么奇怪的UDP重定向了嘛

**Reply by xjasonlyu** [reply]
你用nc连接那个端口是通的吗，UDP也行吗

**Reply by yiguihai11** [reply]
> The `0.0.0.0:0` is not related to this problem, JUST IGNORE IT!
> 
> Why in your environment every port can handle SOCKS5 UDP ASSOCIATE?? That's not expected!
> 
> If the proxy address is not listened, then it should print:
> 
> ```
> error(39, 'Destination address required')
> error(39, 'Destination address required')
> error(39, 'Destination address required')
> error(39, 'Destination address required')
> ...
> ```

你是对的我在termux中测试也是 OSError(89, 'Destination address required') 但是在这台机上确实是pass
我去重装一下系统试下

**Reply by yiguihai11** [reply]
> 你用nc连接那个端口是通的吗，UDP也行吗

不通
![IMG_20220419_170740](https://user-images.githubusercontent.com/103910817/163970347-d2c8be32-90af-453c-ac3f-2aff68477f97.jpg)

**Reply by yiguihai11** [reply]
@zonyitoo @xjasonlyu 找到原因了我 tun2socks 启动参数中的 -interface eth0 去除后就正常了
![IMG_20220420_095521](https://user-images.githubusercontent.com/103910817/164131336-d17a08ad-5457-4e84-b048-219109ca3783.jpg)
![IMG_20220420_095438](https://user-images.githubusercontent.com/103910817/164131347-3a84e411-6b03-44b6-a18e-1116d22f9c69.jpg)
谢谢你们

**Reply by zonyitoo** [reply]
Ok, `created udp association for ...`, so `sslocal` finally receive the UDP request.

**Reply by xjasonlyu** [reply]
> @zonyitoo @xjasonlyu 找到原因了我 tun2socks 启动参数中的 -interface eth0 去除后就正常了

奥等下，你是发到loopback接口的，所以不应该用—interface lo吗（而且发到回环地址是不需要绑定接口的）

**Reply by yiguihai11** [reply]
> > @zonyitoo @xjasonlyu 找到原因了我 tun2socks 启动参数中的 -interface eth0 去除后就正常了
> 
> 奥等下，你是发到loopback接口的，所以不应该用—interface lo吗（而且发到回环地址是不需要绑定接口的）

尴尬！原来是这样....
这个参数是看别人的示例套用的，本着死马当活马医的态度删除后就成功了。

**Comment by sujiacong** [comment]
这个支持badvpn-tun2socks的--udpgw-remote-server-addr 功能吗？

我用ssh socks5功能创建的socks代理不支持socks udp，需要借助badvpn-tun2socksd的这个udp转发功能。
